kong/tools/queue.lua

```diff
@@ -0,0 +1,408 @@
+-- to their new locations
+function Queue.get_params(config)
+  local queue_config = config.queue or {}
+  -- Common queue related legacy parameters
+  if config.retry_count and config.retry_count ~= ngx.null then
+    ngx.log(ngx.WARN, string.format(
+      "deprecated `retry_count` parameter in plugin %s ignored",
+      kong.plugin.get_id()))
+  end
+  if config.queue_size and config.queue_size ~= 1 and config.queue_size ~= ngx.null then
+    ngx.log(ngx.WARN, string.format(
+      "deprecated `queue_size` parameter in plugin %s converted to `queue.batch_max_size`",
+      kong.plugin.get_id()))
+    queue_config.batch_max_size = config.queue_size
+  end
+  if config.flush_timeout and config.flush_timeout ~= 2 and config.flush_timeout ~= ngx.null then
+    ngx.log(ngx.WARN, string.format(
+      "deprecated `flush_timeout` parameter in plugin %s converted to `queue.max_delay`",
+      kong.plugin.get_id()))
+    queue_config.max_delay = config.flush_timeout
+  end
+  -- Queue related opentelemetry plugin parameters
+  if config.batch_span_count and config.batch_span_count ~= 200 and config.batch_span_count ~= ngx.null then
+    ngx.log(ngx.WARN, string.format(
+      "deprecated `batch_span_count` parameter in plugin %s converted to `queue.batch_max_size`",
+      kong.plugin.get_id()))
+    queue_config.batch_max_size = config.batch_span_count
+  end
+  if config.batch_flush_delay and config.batch_flush_delay ~= 3 and config.batch_flush_delay ~= ngx.null then
+    ngx.log(ngx.WARN, string.format(
+      "deprecated `batch_flush_delay` parameter in plugin %s converted to `queue.max_delay`",
+      kong.plugin.get_id()))
+    queue_config.max_delay = config.batch_flush_delay
+  end
+
+  if not queue_config.name then
+    queue_config.name = kong.plugin.get_id()
+  end
+  return queue_config
+end
+
+
+-------------------------------------------------------------------------------
+-- Drain the queue, used during orderly shutdown and for testing.
+-- @param self Queue
+function Queue:_drain()
+end
+
+
+-------------------------------------------------------------------------------
+-- Add entry to the queue
+-- @param conf plugin configuration of the plugin instance that caused the item to be queued
+-- @param entry the value included in the queue. It can be any Lua value besides nil.
+-- @return true, or nil and an error message.
+function Queue:_enqueue(entry)
+  if entry == nil then
+    return nil, "entry must be a non-nil Lua value"
+  end
+
+  if self:count() >= self.capacity * Queue.CAPACITY_WARNING_THRESHOLD then
+    if not self.warned then
+      self:log(WARN, 'queue at %%% capacity', Queue.CAPACITY_WARNING_THRESHOLD * 100)
+      self.warned = true
+    end
+  else
+    self.warned = nil
+  end
+
+  if self:count() == self.capacity then
+    if not self.queue_full then
+      self.queue_full = true
+      self:log(ERR, "queue full, dropping old entries until processing is successful again")
```

Comment:
```suggestion
      self:log(ERR, "queue full, dropping old entries until processing is successful again")
```

based on the state of 'queue_full', is there any change in the behavior of the queue itself?
It seems that when queue is full, new items will remove old items and that's about it.
If that is the case, why even maintain a queue_full?

Reply:
We don't want to drown the log with identical messages.

Reply:
I suggest sampling in that case rather than using a state machine which is a bit confusing. I suggest logging once per minute. We want to catch users' attention while not taking out their logging infrastructure or a huge bill and once a minute seems like a good trade-off to me. What do you think?